File: Python/Recover Binary Search Tree.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None
# O(n)
class Solution:
    # @param root, a tree node
    # @return a tree node
	def recoverTree(self, root):
		self.n1 = None
		self.n2 = None
		self.prev = None
		self.proc(root)
		if self.n1 and self.n2:
			temp = self.n1.val
			self.n1.val = self.n2.val
			self.n2.val = temp
		return root
	def proc(self,root):
		if root == None:
			return
		self.proc(root.left)
		if self.prev == None:
			self.prev = root
		if self.n1 == None and self.prev.val > root.val:
			self.n1 = self.prev
			self.n2 = root
		elif self.prev.val > root.val:
			self.n2 = root
		self.prev = root
		self.proc(root.right)
		return
		
		
class Solution:
    # @param root, a tree node
    # @return a tree node
	def recoverTree(self, root):
		n1 = None
		n2 = None
		prev = None
		self.proc(root,n1,n2,prev)
		if n1 and n2:
			temp = n1.val
			n1.val = n2.val
			n2.val = temp
		return root
	def proc(self,root,n1,n2,prev):
		if root == None:
			return
		self.proc(root.left,n1,n2,prev)
		if prev and prev.val > root.val:
			n2 = root
			if not n1:
				n1 = prev
		prev = root
		self.proc(root.right,n1,n2,prev)
	def change(self,a):
		a = TreeNode(10)

		
a = TreeNode(0)
b = None
s =Solution()
s.change(b)
print b.val

File: Java/Recover Binary Search Tree.java
class Recover Binary Search Tree{
	//left root right
	//two neighbor nodes swap:
	//fron > back
	//non neighbor:
	//fron>back * 2
	public void recoverTree(TreeNode root) {
        ArrayList<TreeNode> res = new ArrayList<TreeNode>();
        ArrayList<TreeNode> pre = new ArrayList<TreeNode>();
        findSwap(root,res,pre);
        if(res.size()!=0){
        	int v = res.get(0).val;
        	res.get(0).val = res.get(1).val;
        	res.get(1).val = v;
        }
    }
    private void findSwap(TreeNode root, ArrayList<TreeNode> res, ArrayList<TreeNode> pre){
    	if(root == null)
    		return;
    	findSwap(root.left,res,pre);
    	if(pre.size() != 0 && root.val < pre.get(0).val){
    		if(res.size() == 0){
    			res.add(pre.get(0));
    			res.add(root);
    		}
    		else{
    			res.set(1,root);
    		}
    	}
    	if(pre.size() == 0)
    		pre.add(root);
    	else
    		pre.set(0,root);
    	findSwap(root.right,res,pre);


    }
}